ctru-rs/src/applets/swkbd.rs

```diff
@@ -675,6 +640,292 @@ impl SoftwareKeyboard {
+            SWKBD_FILTER_CALLBACK, SWKBD_OUTOFMEM,
+        };
+
+        let swkbd = self.state.as_mut();
+        let mut extra = unsafe { swkbd.__bindgen_anon_1.extra };
+
+        // Calculate shared mem size
+        let mut shared_mem_size = 0;
+
+        shared_mem_size +=
+            (std::mem::size_of::<u16>() * (swkbd.max_text_len as usize + 1) + 3) & !3;
+
+        let dict_off = shared_mem_size;
+
+        shared_mem_size +=
+            (std::mem::size_of::<SwkbdDictWord>() * swkbd.dict_word_count as usize + 3) & !3;
+
+        let status_off = shared_mem_size;
+
+        shared_mem_size += if swkbd.initial_learning_offset >= 0 {
+            std::mem::size_of::<SwkbdStatusData>()
+        } else {
+            0
+        };
+
+        let learning_off = shared_mem_size;
+
+        shared_mem_size += if swkbd.initial_learning_offset >= 0 {
+            std::mem::size_of::<SwkbdLearningData>()
+        } else {
+            0
+        };
+
+        if swkbd.save_state_flags & (1 << 0) != 0 {
+            swkbd.status_offset = shared_mem_size as _;
+            shared_mem_size += std::mem::size_of::<SwkbdStatusData>();
+        }
+
+        if swkbd.save_state_flags & (1 << 1) != 0 {
+            swkbd.learning_offset = shared_mem_size as _;
+            shared_mem_size += std::mem::size_of::<SwkbdLearningData>();
+        }
+
+        shared_mem_size = (shared_mem_size + 0xFFF) & !0xFFF;
+
+        swkbd.shared_memory_size = shared_mem_size;
+
+        // Allocate shared mem
+        let swkbd_shared_mem_ptr = unsafe { libc::memalign(0x1000, shared_mem_size) };
+
+        let mut swkbd_shared_mem_handle = 0;
+
+        if swkbd_shared_mem_ptr.is_null() {
+            swkbd.result = SWKBD_OUTOFMEM;
+            return SWKBD_BUTTON_NONE;
+        }
+
+        let res = unsafe {
+            svcCreateMemoryBlock(
+                &mut swkbd_shared_mem_handle,
+                swkbd_shared_mem_ptr as _,
+                shared_mem_size as _,
+                MEMPERM_READ | MEMPERM_WRITE,
+                MEMPERM_READ | MEMPERM_WRITE,
+            )
+        };
+
+        if R_FAILED(res) {
+            unsafe {
+                libc::free(swkbd_shared_mem_ptr);
+                swkbd.result = SWKBD_OUTOFMEM;
+                return SWKBD_BUTTON_NONE;
+            }
+        }
+
+        // Copy stuff to shared mem
+        if !extra.initial_text.is_null() {
+            swkbd.initial_text_offset = 0;
+
+            unsafe {
+                let utf16_iter =
+                    str::from_utf8_unchecked(CStr::from_ptr(extra.initial_text).to_bytes())
+                        .encode_utf16()
+                        .take(swkbd.max_text_len as _)
+                        .chain(once(0));
+
+                let mut initial_text_cursor = swkbd_shared_mem_ptr.cast();
+
+                for code_point in utf16_iter {
+                    *initial_text_cursor = code_point;
+                    initial_text_cursor = initial_text_cursor.add(1);
```

Comment:
Just for my understanding, is this equivalent to e.g.
```rs
let utf16: Vec<u16> = utf16_iter.collect();
std::ptr::copy(utf16.as_ptr(), initial_text_cursor, utf16.len());
```
other than the extra `Vec` allocation? I guess there's no equivalent `ptr::write(impl Iterator)` or something like that

Reply:
Yep, this code is just copying each UTF-16 code unit into the shared memory block.